### DISCUSSION TITLE
Proxmox VE LXC Filesystem Trim

### DISCUSSION THREAD

**Discussion by MickLesk** [community-scripts/ProxmoxVE#805] [Guides]
With traditional hard disk drives (HDDs or spinner disks), there isn't a significant performance penalty for leaving empty blocks or unused space on the disk. This is because HDDs use spinning magnetic platters and mechanical read/write heads to access data. Therefore, the physical location of data on the disk doesn't impact performance significantly, and there's no need to actively manage unused space to maintain performance.

However, solid-state drives (SSDs) operate differently. SSDs use NAND flash memory to store data, and they have a limited number of write cycles per cell. As data is written, erased, and rewritten to the SSD, performance can degrade over time due to the way NAND flash memory works. Specifically, write operations are faster on empty blocks than on blocks that have been previously written and need to be erased before new data can be written.

fstrim is a command used with SSDs to inform the drive which blocks of data are no longer considered in use by the filesystem so that the drive can internally erase them and prepare them for new data. This helps maintain consistent performance on SSDs over time and prevents degradation due to inefficient write operations.

Similarly, thin-provisioned storage technologies like LVM-Thin, Ceph, and qcow2 allocate storage space dynamically as needed. Without proper management, unused blocks within these storage systems can continue to occupy space, causing storage usage to grow unnecessarily.

Therefore, it's essential to periodically trim SSDs and manage unused space in thin-provisioned storage systems to optimize performance and prevent unnecessary consumption of storage resources.

VMs are configured to run fstrim automatically on a weekly basis, for LXC containers, it's necessary to perform fstrim manually or set up a separate automated process.

So, I created a script to run fstrim on all LXC containers (that are not selected to be skipped). This is designed to work with SSDs on ext4 filesystems only.

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/community-scripts/ProxmoxVE/main/tools/pve/fstrim.sh)""
```

### Migrated from: https://github.com/tteck/Proxmox/discussions/2505

**Comment by vgdh** [comment]
why run trim on lxc? Isn't it better to run trim on the host?

**Comment by djismgaming** [comment]
How could we automate this script by having an option to run it without confirmation?

**Reply by Citroene** [reply]
This!

i want to run the script with cron

**Reply by towerhand** [reply]
I use:

0 1 * * 6 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /bin/bash -c 'pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null'

**Reply by djismgaming** [reply]
> I use:
> 
> 0 1 * * 6 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /bin/bash -c 'pct list | awk "/^[0-9]/ {print $1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null'

Awesome! Gave me a great start.

I think your command is just missing a slash: `...{print \$1}...` instead of `...{print $1}...`

Full working command:
```bash
pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null
```

Full working `cron` entry (running at 1:00 AM every Saturday in this example): 
```bash
0 1 * * 6 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /bin/bash -c 'pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null'
```

**Reply by Citroene** [reply]
Thx! 

is there also a way to use the LXC clean script with cron?
`bash -c "$(wget -qLO - https://github.com/community-scripts/ProxmoxVE/raw/main/misc/clean-lxcs.sh)"`

**Reply by Patrick010** [reply]
Sweet! Never ran pct fstrim, it returned me 231Gb :)
Added to my cron immediately!

I had ChatGTP check the 'Full working cron entry' and it complained about backticks. Below a supposedly improved cron entry.

`0 1 * * 6 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /bin/bash -c 'pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim "$ct"; done >> /var/log/fstrim-cron.log 2>&1'`

**Reply by djismgaming** [reply]
You're completely right @Patrick010 thanks for pointing that out! I updated my entry now with the replacement of backticks..

**Reply by ncodee** [reply]
> > I use:
> > 0 1 * * 6 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /bin/bash -c 'pct list | awk "/^[0-9]/ {print $1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null'
> 
> Awesome! Gave me a great start.
> 
> I think your command is just missing a slash: `...{print \$1}...` instead of `...{print $1}...`
> 
> Full working command:
> 
> ```shell
> pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null
> ```
> 
> Full working `cron` entry (running at 1:00 AM every Saturday in this example):
> 
> ```shell
> 0 1 * * 6 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /bin/bash -c 'pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim $ct; done >>/var/log/fstrim-cron.log 2>/dev/null'
> ```

I'd gather that this runs on all containers? Possible to skip `stopped` containers? thanks

**Reply by djismgaming** [reply]
Does it do anything weird or bad to the stopped containers? I don't think so. Space that does not need trimming will not be re-trimmed IIRC. Trim is just marking space as free.

**Reply by ncodee** [reply]
> Does it do anything weird or bad to the stopped containers? I don't think so. Space that does not need trimming will not be re-trimmed IIRC. Trim is just marking space as free.

That's not the issue. In order for the stopped containers to be trimmed, they will need to be running. If the cron doesn't respect the 'skipping', then all stopped containers will restart and then stop again after it has trimmed, which has a load on the CPU.

**Reply by djismgaming** [reply]
I just tried manually trimming a container with `pct fstrim 399` and it does lock the container configuration for a second but did not start the container.

![unknown_replay_2025 06 16-00 19_1](https://github.com/user-attachments/assets/cf3cf458-9d69-4ea2-b559-59d30c69ce76)

**Comment by yggdrasil75** [comment]
this appears to have been moved or deleted.

**Reply by tremor021** [reply]
its still there, we are just migrating our website to new frontend. will be fixed shortly

**Reply by tremor021** [reply]
`bash -c "$(curl -fsSL https://raw.githubusercontent.com/community-scripts/ProxmoxVE/main/tools/pve/fstrim.sh)"`
